snorkel/slicing/utils.py

```diff
@@ -0,0 +1,202 @@
+
+from .modules.slice_combiner import SliceCombinerModule
+
+
+def add_slice_labels(
+    base_task: Task,
+    dataloader: MultitaskDataLoader,
+    slice_labels: csr_matrix,
+    slice_names: List[str],
+):
+    slice_labels = slice_labels.toarray()
+    slice_labels, slice_names = _add_base_slice(slice_labels, slice_names)
+
+    label_name = dataloader.task_to_label_dict[base_task.name]
+    labels = dataloader.dataset.Y_dict[label_name]
+    for i, slice_name in enumerate(slice_names):
+
+        # Convert labels
+        indicators = torch.LongTensor(slice_labels[:, i])
+        ind_labels = convert_labels(indicators, source="onezero", target="categorical")
+        pred_labels = indicators * labels
+
+        ind_task_name = f"{base_task.name}_slice:{slice_name}_ind"
+        pred_task_name = f"{base_task.name}_slice:{slice_name}_pred"
+
+        # Update dataloaders
+        dataloader.dataset.Y_dict[ind_task_name] = ind_labels
+        dataloader.dataset.Y_dict[pred_task_name] = pred_labels
+
+        dataloader.task_to_label_dict[ind_task_name] = ind_task_name
+        dataloader.task_to_label_dict[pred_task_name] = pred_task_name
+    return dataloader
+
+
+def convert_to_slice_tasks(base_task: Task, slice_names: List[str]):
+    """Adds slice labels to dataloader and creates new slice tasks (including base slice)
+
+    Each slice will get two slice-specific heads:
+    - an indicator head that learns to identify when DataPoints are in that slice
+    - a predictor head that is trained on only members of that slice
+
+    The base task's head is replaced by a master head that makes predictions based on
+    a combination of the predictor heads' predictions that are weighted by the
+    indicator heads' predictions.
+
+    NOTE: The current implementation pollutes the module_pool---the indicator's task's
+    module_pool includes predictor modules and vice versa since both are modified in
+    place. This does not affect the result because the task flows dictate which modules
+    get used, and those do not include the extra modules. An alternative would be to
+    make separate copies of the module pool for each, but that wastes time and memory
+    extra copies of (potentially very large) modules that will be merged in a moment
+    away in the model since they have the same name. We leave resolution of this issue
+    for a future release.
+    """
+    if "base" not in slice_names:
+        slice_names = slice_names + ["base"]
+
+    tasks = []
+
+    # NOTE: We assume here that the last operation uses the head module
+    # Identify base task head module
+    head_module_op = base_task.task_flow[-1]
+    head_module = base_task.module_pool[head_module_op.module_name]
+
+    if isinstance(head_module, nn.DataParallel):
+        head_module = head_module.module
+
+    neck_size = head_module.in_features
+    base_task_cardinality = head_module.out_features
+
+    # Remove the slice-unaware head module from module pool and task flow
```

Comment:
I got confused about this when I read line 101—a quick glance made think the final prediction head was still intact. 

could we rename the base task flow? maybe as `body_flow` or something to indicate we have decapitated the original task